packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -0,0 +1,209 @@
+///         features necessary for secure transfers ERC20 tokens between L2 chains. Messages sent through the
+///         L2ToL2CrossDomainMessenger on the source chain receive both replay protection as well as domain binding.
+contract L2ToL2CrossDomainMessenger is IL2ToL2CrossDomainMessenger, ISemver, ReentrancyGuard {
+    /// @notice Storage slot for `entered` value.
+    ///         Equal to bytes32(uint256(keccak256("l2tol2crossdomainmessenger.entered")) - 1)
+    bytes32 internal constant ENTERED_SLOT = 0xf53fc38c5e461bdcbbeb47887fecf014abd399293109cd50f65e5f9078cfd025;
+
+    /// @notice Storage slot for the sender of the current cross domain message.
+    ///         Equal to bytes32(uint256(keccak256("l2tol2crossdomainmessenger.sender")) - 1)
+    bytes32 internal constant CROSS_DOMAIN_MESSAGE_SENDER_SLOT =
+        0xb83444d07072b122e2e72a669ce32857d892345c19856f4e7142d06a167ab3f3;
+
+    /// @notice Storage slot for the source of the current cross domain message.
+    ///         Equal to bytes32(uint256(keccak256("l2tol2crossdomainmessenger.source")) - 1)
+    bytes32 internal constant CROSS_DOMAIN_MESSAGE_SOURCE_SLOT =
+        0x711dfa3259c842fffc17d6e1f1e0fc5927756133a2345ca56b4cb8178589fee7;
+
+    /// @notice Current message version identifier.
+    uint16 public constant messageVersion = uint16(0);
+
+    /// @notice Semantic version.
+    /// @custom:semver 1.0.0
+    string public constant version = "1.0.0";
+
+    /// @notice Mapping of message hashes to boolean receipt values. Note that a message will only be present in this
+    ///         mapping if it has successfully been relayed on this chain, and can therefore not be relayed again.
+    mapping(bytes32 => bool) public successfulMessages;
+
+    /// @notice Nonce for the next message to be sent, without the message version applied. Use the messageNonce getter,
+    ///         which will insert the message version into the nonce to give you the actual nonce to be used for the
+    ///         message.
+    uint240 internal msgNonce;
+
+    /// @notice Emitted whenever a message is sent to the other chain.
+    /// @param data Encoded data of the message that was sent.
+    event SentMessage(bytes data) anonymous;
+
+    /// @notice Emitted whenever a message is successfully relayed on this chain.
+    /// @param messageHash Hash of the message that was relayed.
+    event RelayedMessage(bytes32 indexed messageHash);
+
+    /// @notice Emitted whenever a message fails to be relayed on this chain.
+    /// @param messageHash Hash of the message that failed to be relayed.
+    event FailedRelayedMessage(bytes32 indexed messageHash);
+
+    /// @notice Enforces that cross domain message sender and source are set. Reverts if not.
+    ///         Used to differentiate between 0 and nil in transient storage.
+    modifier notEntered() {
+        if (TransientContext.get(ENTERED_SLOT) == 0) revert NotEntered();
```

Comment:
This is doing extra work with hashing together the entered slot value, you could use `TransientContext` like this:

````solidity
if (TransientContext.callDepth() != 0) revert NonReentrant();
TransientContext.increment();
_;
TransientContext.decrement();
````

Reply:
the problem is that we replaced `reentrancyAware` by `nonReentrant`, so the depth is always we zero